clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java

```diff
@@ -82,6 +92,9 @@ public long readFromReadableChannel(ReadableByteChannel channel) throws IOExcept
             if (bytesRead < 0)
                 throw new EOFException();
             read += bytesRead;
+
+            ensureNotHandshakeFailurePacket(channel, size);
```

Comment:
This is an expensive check. It needs to be done only at the start of a connection, not for every receive. I think it would be better to trigger the check from one of the channel/transport layer classes.